### PR TITLE
Add a safe way to grab the project GUID

### PR DESCRIPTION
**Pull request by davkean**
The project GUID can be set by the solution right up to the point it fires IVsSolutionEvents.OnAfterOpenProject (IProjectAsyncLoadDashboard.ProjectLoadedInHost completes) in reaction to a clash with another project. Add a singleton API that is intended as a wrapper around IProjectGuidService/IProjectGuidService2 that waits for the project to be loaded before retrieving the GUID..

Fixes: #3338

Updated services to consume that.

Still testing this, but code is complete.

### CODE REVIEW

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/VsSafeProjectGuidService.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.ComponentModel.Composition;
+using System.Threading.Tasks;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS
+{
+    /// <summary>
+    ///     An implementation of <see cref="IProjectGuidSafeAccessor"/> that waits until the project
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/VsSafeProjectGuidService.cs]: Ugh, old name need to fix that.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/VsSafeProjectGuidServiceTests.cs [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+namespace Microsoft.VisualStudio.ProjectSystem.VS
+{
+    [Trait("UnitTest", "ProjectSystem")]
+    public class VsSafeProjectGuidServiceTests
+    {
+        [Fact]
+        public void GetProjectGuidAsync_WhenProjectAlreadyUnloaded_ReturnsCancelledTask()
+        {
+            var tasksService = IProjectAsynchronousTasksServiceFactory.ImplementUnloadCancellationToken(new CancellationToken(canceled: true));
+            var loadDashboard = IProjectAsyncLoadDashboardFactory.ImplementProjectLoadedInHost(() => Task.Delay(-1));
```
Comment (Pilchie) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/VsSafeProjectGuidServiceTests.cs]: What does `-1` mean here?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Debug/StartupProjectRegistrar.cs [hunk 3]
```diff
@@ -69,12 +73,8 @@ public Task InitializeAsync()
 
         protected override async Task InitializeCoreAsync(CancellationToken cancellationToken)
         {
-            IProjectGuidService2 projectGuidService = ProjectGuidServices.FirstOrDefault()?.Value as IProjectGuidService2;
```
Comment (Pilchie) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Debug/StartupProjectRegistrar.cs]: Should `ProjectGuidServices` be removed?

Comment (davkean) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Debug/StartupProjectRegistrar.cs]: Yes it should.

**Comment by davkean** [comment]
> Is there any concern about these async calls being delayed until the project is loaded? 
No. Language service already waited for ProjectLoadedInHost. 

>Will this significantly affect PPL or introduce more UI delays immediately after the solution is loaded?
No, Language service behavior is exactly the same. The StartupProjectRegistrar behavior has slightly changed in that it runs after open - but I removed it switch to the UI thread a while back.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/VsSafeProjectGuidServiceTests.cs (hunk 2 above)
Comment (davkean) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/VsSafeProjectGuidServiceTests.cs]: Infinite.

Comment (davkean) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/VsSafeProjectGuidServiceTests.cs]: Will fix.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Debug/StartupProjectRegistrar.cs (hunk 3 above)
Comment (davkean) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Debug/StartupProjectRegistrar.cs]: Fixed.

**Comment by davkean** [comment]
This looks good.

**Pull request merged**